### ISSUE TITLE
Gaussian NLL loss

### ISSUE THREAD

**Issue by nailimixaM** [pytorch/pytorch#48520]
## 🚀 Feature
Gaussian negative log-likelihood loss, similar to issue #1774 (and solution pull #1779) 

## Motivation

The homoscedastic Gaussian loss is described in Equation 1 of this [paper](https://arxiv.org/pdf/1612.01474.pdf). The heteroscedastic version in Equation 2 [here](https://arxiv.org/pdf/2011.02838.pdf)  (ignoring the final anchoring loss term). These are both key to the uncertainty quantification techniques described. 

## Pitch

I'm happy to implement this, using the template of pull #1779. The implementation will allow for both homoscedastic and heteroscedastic losses.  

## Alternatives

An alternative would be to instantiate a Gaussian (https://pytorch.org/docs/stable/distributions.html#normal) and evaluate the log of this. However, this seems wasteful given a new Gaussian would be instantiated for every new function call for the best case (homoscedastic), and for every element of the output-target pairs in the worst case (heteroscedastic). 

## Additional context

Definitions: [homo/heteroscedasticity](https://en.wikipedia.org/wiki/Homoscedasticity) (wiki)

cc @albanD @mruberry

**Comment by nailimixaM** [comment]
cc @fmassa @ezyang

**Comment by sparkingdark** [comment]
Hey @nailimixaM I am happy to collaborate to implement this functionality,let me know if I can help.

@fmassa @ezyang

**Comment by albanD** [comment]
Hi,

Thanks for opening a feature request!

The general policy we have for new functions/layers is to make sure they have been used by researchers significantly since their publication (the main reason being that once something is added to torch.nn, it cannot really be removed anymore and we need to maintain it).

So we can look if the homoscedastic version meet that criterion (I haven't looked at the paper/citations/repos that use it in details, we can discuss this further below if you want) but I think the heteroscedastic version does not for sure as it was just published.
We usually recommend that people provide these new functions in their own repo to begin with and we can reconsider this in 6 months.
If you need anything from the core lib to be able to implement it in your own repo, don't hesitate to ask and we'll be very happy to help!

**Comment by sparkingdark** [comment]
hey @albanD can I also give it a try to implement in my repo?

**Comment by nailimixaM** [comment]
Thanks @albanD, I understand your point. I realise I've made a slight slip-up - the first paper proposes the heteroscedastic loss (which is a general version of the homoscedastic loss)! (The homoscedastic loss is simply the case where the variances are all equal, which is trivially handled by a heteroscedastic implementation). In this case I'd be interested to hear whether you think this loss qualifies, given what I understand to be the high-impact nature of the paper (1k+ citations on Google scholar) and the widely used loss function.

> If you need anything from the core lib to be able to implement it in your own repo, don't hesitate to ask and we'll be very happy to help!

Thanks, may take you up on this!

@sparkingdark Happy to collaborate if this gets the green light!

**Comment by nailimixaM** [comment]
@albanD See for example this tensorflow commit: https://github.com/tensorflow/agents/commit/3ae8ef2a8eb0c8496998e269159a3d20e36d5f45

**Comment by sparkingdark** [comment]
Thanks man I love to do it ...

**Comment by albanD** [comment]
Hey,

That sounds good, I think we would be happy to add a Module and functional for the general case.
Could you share here what you think the API will look like just to make sure everyone is happy with that before you start implementing it?

**Comment by nailimixaM** [comment]
Great! Sure, I had the following in mind, trying to keep the same theme as the [Poisson NLL loss]:(https://pytorch.org/docs/stable/generated/torch.nn.PoissonNLLLoss.html#torch.nn.PoissonNLLLoss)

`
def gaussian_nll_loss(input, target, var, eps=1e-8, full=False, reduction='mean'):
        r"""Gaussian negative log likelihood loss.

        See :class:`~torch.nn.GaussianNLLLoss` for details.

        Args:
                input: tensor of shape=(N, *).
                target: tensor with same shape as input.
                var: tensor of positive variance(s), shape = (N, 1) or
                        same shape as input.
                eps: value added to var, for stability
                full: True/False (bool), include the constant term in the loss
                        calculation. Default: False.
                reduction: specifies the reduction to apply to the output:
                        'none' | 'mean' | 'sum'. 'none': no reduction will be
                        applied, 'mean': the sum of the output will be divided
                        by the number of elements in the output, 'sum': the
                        output will be summed. Default: 'mean'.

        Output:
                loss: scalar by default. If reduction is 'none', then (N, *),
                        same shape as input.
        """`

**Comment by albanD** [comment]
That looks good!
We will want a bit more details for things like eps but we can discuss that on the PR.

Let me know if you want any guidelines on how to add this or have any question! You can just add me as a reviewer on the PR once you have it ready!

**Comment by Medaminemairech** [comment]
Hi, can you help me implement something similar for a more general Gaussian Mixture Distribution?

**Comment by nailimixaM** [comment]
Hi @Medaminemairech, if you are looking to calculate the negative log likelihood of a pdf which can be described as a sum of K Gaussians with prior weights `phi = [phi_0, phi_1 ... phi_K-1]`  and where you are trying to learn the means and variances of each Gaussian, then you could try something similar to below, which is for 2 Gaussians in 3D:

```
phi = [0.4, 0.6]  # prior weights
inputs = [torch.tensor([0., 0., 0.], requires_grad=True), torch.tensor([1., 1., 1.], requires_grad=True)] # predicted means
vars = [torch.tensor([0.3, 0.4, 0.5], requires_grad=True), torch.tensor([.8, 0.1, .4], requires_grad=True)] # predicted variances
target = torch.tensor([1., 0., -1.])  # sample from pdf

GaussianNLLLoss = torch.nn.GaussianNLLLoss()
total_likelihood = 0

# Calculate and sum the likelihoods of each component
for i, phi_i in enumerate(phi):
  loss = GaussianNLLLoss(inputs[i], target, vars[i])
  likelihood = torch.exp(-loss)
  total_likelihood += phi_i * likelihood

# Take the negative log to give the GMM negative log-likelihood loss
GMM_NLLLoss = -torch.log(total_likelihood) 

print(GMM_NLLLoss) # Gives tensor(1.2013, grad_fn=<NegBackward0>)
```

Let me know if that helps!

**Comment by Medaminemairech** [comment]
HI @nailimixaM, yep the weighted sum should do the job. Thanks a lot! 😄
